extension/bearertokenauthextension/bearertokenauth.go

```diff
@@ -44,26 +47,99 @@ func (c *PerRPCAuth) RequireTransportSecurity() bool {
 
 // BearerTokenAuth is an implementation of configauth.GRPCClientAuthenticator. It embeds a static authorization "bearer" token in every rpc call.
 type BearerTokenAuth struct {
-	tokenString string
-	logger      *zap.Logger
+	muTokenString sync.RWMutex
+	tokenString   string
+
+	shutdownCH chan struct{}
+
+	tokenFilename string
+	logger        *zap.Logger
 }
 
 var _ configauth.ClientAuthenticator = (*BearerTokenAuth)(nil)
 
 func newBearerTokenAuth(cfg *Config, logger *zap.Logger) *BearerTokenAuth {
+	if cfg.BearerTokenFilename != "" && cfg.BearerToken != "" {
+		logger.Warn("a filename is specified. Configured token is ignored!")
+	}
 	return &BearerTokenAuth{
-		tokenString: cfg.BearerToken,
-		logger:      logger,
+		tokenString:   cfg.BearerToken,
+		tokenFilename: cfg.BearerTokenFilename,
+		logger:        logger,
 	}
 }
 
-// Start of BearerTokenAuth does nothing and returns nil
+// Start of BearerTokenAuth does nothing and returns nil if no tokenFilename
+// is specified. Otherwise a routine is started to monitor the file containing
+// the token to be transferred.
 func (b *BearerTokenAuth) Start(ctx context.Context, host component.Host) error {
-	return nil
+	if b.tokenFilename == "" {
+		return nil
+	}
+
+	if b.shutdownCH != nil {
+		return fmt.Errorf("bearerToken file monitoring is already running")
+	}
+
+	// Read file once
+	tokenStr, err := os.ReadFile(b.tokenFilename)
+	if err != nil {
+		return err
+	}
+	b.muTokenString.Lock()
+	b.tokenString = string(tokenStr)
+	b.muTokenString.Unlock()
+
+	b.shutdownCH = make(chan struct{})
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+
+	// start file watcher
+	go func(ctx context.Context) {
+		defer watcher.Close()
+		for {
+			select {
+			case _, ok := <-b.shutdownCH:
+				_ = ok
+				return
+			case <-ctx.Done():
+				return
+			case event, ok := <-watcher.Events:
+				if !ok {
+					continue
+				}
+				if event.Op == fsnotify.Write {
```

Comment:
Is this the only event we are interested in? I remember dealing with a similar code in Kubernetes and the file we were watching was a configmap (not a secret), which has a few levels of links before getting to the actual file, meaning that the event was never "write" on the link we were watching.

I guess my real question is: have you tried this with an actual secret on a Kubernetes cluster?

Reply:
Actually only locally. Let me verify configmaps and secrets.

Reply:
Seems we need to react to `CHMOD` and `REMOVE` events too. Then its working :) 
<details>
  <summary>Manifests used to test</summary>
```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: otlp-config
data:
  config.yaml: |
    extensions:
      bearertokenauth:
        filename: /app/secert/token.txt

    receivers:
      otlp:
        protocols:
          http:

    exporters:
      logging:
      otlphttp/withauth:
        endpoint: http://localhost:9000
        auth:
          authenticator: bearertokenauth

    service:
      extensions: [bearertokenauth]
      pipelines:
        traces:
          receivers: [otlp]
          exporters: [otlphttp/withauth]
---
apiVersion: v1
kind: Secret
metadata:
  name: otlp-token
type: Opaque
data: # test values 1. test123 2. test456
  #token.txt: dGVzdDEyMwo=
  token.txt: dGVzdDQ1Ngo=
---
apiVersion: v1
kind: Pod
metadata:
  name: mypod
spec:
  containers:
    - name: otlp-test
      image: ghcr.io/frzifus/otelcontribcol:print-7
      args: ["--config", "/app/config.yaml"]
      volumeMounts:
        - name: token-volume
          mountPath: /app/secert
          readOnly: true
        - name: config-volume
          mountPath: /app/
          readOnly: true
  volumes:
    - name: token-volume
      secret:
        secretName: otlp-token
    - name: config-volume
      configMap:
        name: otlp-config
```
</details>